src/Stackage/Package/IndexConduit.hs

```diff
@@ -48,30 +44,80 @@ import Distribution.PackageDescription.Parse
 import Distribution.Text (disp, parse)
 import Distribution.Version (VersionRange)
 import qualified Distribution.Text
-import System.FilePath ((</>), (<.>))
-import System.IO (IOMode(ReadMode), hClose, openBinaryFile)
+import Network.HTTP.Client.Conduit
 import Text.PrettyPrint (render)
 
-sourceTarFile
-  :: MonadResource m
-  => Bool -> FilePath -> Producer m Tar.Entry
-sourceTarFile toUngzip fp = do
-  bracketP (openBinaryFile fp ReadMode) hClose $
-    \h -> do
-      lbs <- liftIO $ L.hGetContents h
-      sourceEntries $ Tar.read $ ungzip' lbs
-  where
-    ungzip'
-      | toUngzip = decompress
-      | otherwise = id
+#if MIN_VERSION_http_conduit(2,2,1)
+import Network.HTTP.Simple (httpSource)
+#else
+import qualified Network.HTTP.Client as H
+import qualified Network.HTTP.Client.TLS as H
+
+-- | For future compatibility, copied here from http-conduit-2.2.1
+httpSource
+  :: (MonadResource m, MonadIO n)
+  => Request
+  -> (Response (ConduitM i ByteString n ()) -> ConduitM i o m r)
+  -> ConduitM i o m r
+httpSource req withRes = do
+  man <- liftIO H.getGlobalManager
+  bracketP
+    (H.responseOpen req man)
+    H.responseClose
+    (withRes . fmap bodyReaderSource)
+#endif
+
+data StopException = StopException deriving Show
+
+instance Exception StopException
+
+-- | Download a tarball from a webserver, decompress, parse it and handle it
+-- using a provided `Sink`. Using a conditional function it is possible to
+-- prevent a tarball from being downloaded, for instance in such a case when an
+-- unexpected response status was received, in which case `Nothing` will be
+-- returned. That function also allows to return any value that depends on a
+-- `Response`.
+httpTarballSink
+  :: (MonadActive m, MonadCatch m, MonadResource m, MonadBaseControl IO m)
+  => Request -- ^ Request to the tarball file.
+  -> Bool -- ^ Is the tarball gzipped?
+  -> Sink Tar.Entry m b -- ^ The sink of how entries in the tar file should be
```

Comment:
I need to be able to return a value that depends on a `Response`, sometimes with / sometimes without processing the body, in particular it is used in `Main` to return `etag`. So just having `(Response () -> Sink Tar.Entry m a)` wouldn't work, this is because there is also `Maybe b`. That's why the return of the function is `m (Maybe a, Maybe b)`, there are indeed four scenarios for return values.